### PR TITLE
net: tcp2: Fix sending to 6lo based networks

### LINKED ISSUE #29771: samples: net: sockets: tcp: tcp2 server not accepting with ipv6 bsd sockets
**Describe the bug**
Prior to 70dae094ba2dd3e1f02684a8d673d434966b8b76 it was possible to run an IPv6 TCP server using the default TCP implementation and the BSD sockets API over IEEE 802.15.4. However, since TCP2 became the default TCP implementation, it has not been possible (with the same configuration).

Furthermore, it is unclear if and when it was actually possible to do so (with the same configuration). I went back as far as I could compile in a `git bisect` with TCP2 enabled by default and this configuration has never seemed to work.

It's not clear to me if the problem only affects applications that use the BSD sockets API, but that is how I noticed.

**To Reproduce**

1. apply the [attached patch](https://github.com/zephyrproject-rtos/zephyr/files/5485176/ieee802154-ipv6.patch.txt) for the `samples/net/sockets/echo` application (adds IPv6 / 15.4)
2. select a board that is supports IEEE 802.15.4 (e.g. nrf52840dk_nrf52840)
3. `west build -b `nrf52840dk_nrf52840` -t flash samples/net/sockets/echo -- -DOVERLAY_CONFIG=overlay-802154.conf`
4. telnet to `2001:db8::1` and try to see if echo works (it does not even connect for me.
5. adjust samples/net/sockets/echo/overlay-802154.conf and swap the values for `CONFIG_NET_TCP1=n` and `CONFIG_NET_TCP2=y` to enable `TCP1`.
6. repeat from step 3

I've tried this using `nrf52840dk_nrf52840`, `cc1352r1_launchxl`, and the `cc1352r_sensortag` using IEEE 802.15.4. I've not tried with BLE / 6lowpan yet, but would be surprised if it were unaffected.

**Expected behavior**
The server should echo each line of text back to the client.

**Impact**
While it's possible to explicitly select `CONFIG_NET_TCP1`, it's probably something that should be fixed for TCP2 given that it's the default.

**Logs and console output**
With TCP2 (`telnet` never connected):
```bash
$ ping6 2001:db8::1
PING 2001:db8::1(2001:db8::1) 56 data bytes
64 bytes from 2001:db8::1: icmp_seq=1 ttl=64 time=27.0 ms
64 bytes from 2001:db8::1: icmp_seq=2 ttl=64 time=16.1 ms
64 bytes from 2001:db8::1: icmp_seq=3 ttl=64 time=13.8 ms
^C
--- 2001:db8::1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 13.807/18.974/27.021/5.766 ms
cfriedt@cfriedt-MacBookPro:~/workspace/zephyrproject/zephyr$ telnet 2001:db8::1 4242
Trying 2001:db8::1...
^C
```

With TCP1:
```bash
$ ping6 2001:db8::1
PING 2001:db8::1(2001:db8::1) 56 data bytes
64 bytes from 2001:db8::1: icmp_seq=1 ttl=64 time=12.4 ms
64 bytes from 2001:db8::1: icmp_seq=2 ttl=64 time=13.2 ms
^C
--- 2001:db8::1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1002ms
rtt min/avg/max/mdev = 12.412/12.802/13.193/0.390 ms
cfriedt@cfriedt-MacBookPro:~/workspace/zephyrproject/zephyr$ telnet 2001:db8::1 4242
Trying 2001:db8::1...
Connected to 2001:db8::1.
Escape character is '^]'.
hello?
hello?
```

**Environment (please complete the following information):**
 - OS: Ubuntu Linux (Focal Fossa)
 - Toolchain: Zephyr SDK 0.11.3
 - Commit SHA: 761f1c4bedc4fc617920949c168f75b9d148badc

**Additional context**
Seems to be related to #28577

cc @jukkar @MaureenHelm

### PR DESCRIPTION
**Pull request by jukkar**
Do not send the original pkt in 6lo based networks as in those
the IPv6 header is mangled and we would not be able to do any
resends of the original pkt. So for 6lo networks, clone the
pkt and send it to peer. The original pkt is kept in sent list
in case we need to resend to peer.

Fixes #29771

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by jukkar** [comment]
@cfriedt I have a fix proposal here, I tested it with 802.15.4 simulator and not with real hw. It would be great if you are able to test this with your setup and give info does it work for you too.

**Comment by cfriedt** [comment]
> @cfriedt I have a fix proposal here, I tested it with 802.15.4 simulator and not with real hw. It would be great if you are able to test this with your setup and give info does it work for you too.

Will do - give me 5 min!

**Comment by cfriedt** [comment]
> > @cfriedt I have a fix proposal here, I tested it with 802.15.4 simulator and not with real hw. It would be great if you are able to test this with your setup and give info does it work for you too.
> 
> Will do - give me 5 min!
Did not do the trick unfortunately :(

**Comment by jukkar** [comment]
> Did not do the trick unfortunately :(

Hmm, that is unfortunate, it worked for me. I wonder what is wrong now :(

**Comment by cfriedt** [comment]
> Hmm, that is unfortunate, it worked for me. I wonder what is wrong now :(

It's odd. I wonder if there is something more like an integration test case that could be introduced to spin up qemu and then use python to connect to the echo port. Should work in all test environments. I'm just not sure about the qemu part.

**Comment by jukkar** [comment]
I tested this PR like this:
Terminal 1:
```
west build -b qemu_x86 -d build/server samples/net/sockets/echo_server -- -G'Unix Makefiles' -DOVERLAY_CONFIG=overlay-qemu_802154.conf -DPCAP=capture.pcap -DCONFIG_NET_TCP=y
west build -t server -d build/server
```
Terminal 2:
```
west build -b qemu_x86 -d build/client samples/net/sockets/echo_client -- -DOVERLAY_CONFIG=overlay-qemu_802154.conf -DCONFIG_NET_TCP=y
west build -t client -d build/client
```
See data being transferred between the two QEMUs for some time. Stop QEMUs. Then open `build/server/capture.pcap` file and verify that data is properly set in the packets.

**Comment by cfriedt** [comment]
> I tested this PR like this:
> Terminal 1:
> 
> ```
> west build -b qemu_x86 -d build/server samples/net/sockets/echo_server -- -G'Unix Makefiles' -DOVERLAY_CONFIG=overlay-qemu_802154.conf -DPCAP=capture.pcap -DCONFIG_NET_TCP=y
> west build -t server -d build/server
> ```
> 
> Terminal 2:
> 
> ```
> west build -b qemu_x86 -d build/client samples/net/sockets/echo_client -- -DOVERLAY_CONFIG=overlay-qemu_802154.conf -DCONFIG_NET_TCP=y
> west build -t client -d build/client
> ```
> 
> See data being transferred between the two QEMUs for some time. Stop QEMUs. Then open `build/server/capture.pcap` file and verify that data is properly set in the packets.

Thanks - I'll see if I can reproduce this. I wonder if it has soemething to do with the ARM architecture being used in my case. Soemetimes there may be slight variations in behaviour due to alignment requirements.

**Comment by cfriedt** [comment]
> I tested this PR like this:
> Terminal 1:
> 
> ```
> west build -b qemu_x86 -d build/server samples/net/sockets/echo_server -- -G'Unix Makefiles' -DOVERLAY_CONFIG=overlay-qemu_802154.conf -DPCAP=capture.pcap -DCONFIG_NET_TCP=y
> west build -t server -d build/server
> ```
> 
> Terminal 2:
> 
> ```
> west build -b qemu_x86 -d build/client samples/net/sockets/echo_client -- -DOVERLAY_CONFIG=overlay-qemu_802154.conf -DCONFIG_NET_TCP=y
> west build -t client -d build/client
> ```
> 
> See data being transferred between the two QEMUs for some time. Stop QEMUs. Then open `build/server/capture.pcap` file and verify that data is properly set in the packets.

Is there an additional step needed, @jukkar? Maybe something in `net-tools`? I'm not seeing any packets transferred.

**Comment by jukkar** [comment]
I have the latest from net-tools, commit a265561ad7fabe8539d171116d8f8563cdf5d3ba

I am not seeing that error. If I do not have this PR applied, there are some TCP data sent by the client but whenever there the data needs to be re-transmitted, the connection will fail. If this PR is applied, the connection works as expected.

**Comment by cfriedt** [comment]
Let me try again

**Comment by cfriedt** [comment]
@jukkar - confirmed working with `cc1352r1_launchxl` and `-DOVERLAY_CONFIG=overlay-qemu_802154.conf -DCONFIG_NET_TCP=y`. Thanks for your quick response on this one!

**Review by cfriedt — APPROVED** [review]
Tested with `cc1352r1_launchxl` and `overlay-802154.conf`.

**Review by rlubos — APPROVED** [review]
